File: python/tests/test_fastq.py
"""Tests for FASTQ parsing functionality."""

import bz2
import gzip
import tempfile
from pathlib import Path

import pytest

from prseq.fastq import FastqReader, FastqRecord, read_fastq


def create_test_fastq() -> Path:
    """Create a temporary FASTQ file for testing."""
    content = """@seq1 test sequence
ATCGGATCCTAG
+
IIIIIIIIIIII
@seq2 another test
GGCCTTAAGGGG
+seq2 another test
JJJJJJJJJJJJ
@seq3 short
ATCG
+
AAAA
"""
    temp_file = tempfile.NamedTemporaryFile(mode='w', suffix='.fastq', delete=False)
    temp_file.write(content)
    temp_file.close()
    return Path(temp_file.name)


def create_compressed_test_fastq(compression: str = 'gzip') -> Path:
    """Create a temporary compressed FASTQ file for testing."""
    content = b"""@seq1 compressed
ATCGGATCC
+
IIIIIIIII
@seq2 compressed too
GGCCTTAA
+
JJJJJJJJ
"""

    if compression == 'gzip':
        compressed_content = gzip.compress(content)
        suffix = '.fastq.gz'
    elif compression == 'bzip2':
        compressed_content = bz2.compress(content)
        suffix = '.fastq.bz2'
    else:
        raise ValueError(f"Unsupported compression: {compression}")

    temp_file = tempfile.NamedTemporaryFile(suffix=suffix, delete=False)
    temp_file.write(compressed_content)
    temp_file.close()
    return Path(temp_file.name)


def test_fastq_reader_iterator() -> None:
    """Test iterating over FASTQ records."""
    fastq_file = create_test_fastq()
    try:
        reader = FastqReader.from_file(str(fastq_file))
        records: list[FastqRecord] = list(reader)

        assert len(records) == 3

        # Test first record
        record1 = records[0]
        assert record1.id == "seq1 test sequence"
        assert record1.sequence == "ATCGGATCCTAG"
        assert record1.quality == "IIIIIIIIIIII"

        # Test second record with ID in '+' line
        record2 = records[1]
        assert record2.id == "seq2 another test"
        assert record2.sequence == "GGCCTTAAGGGG"
        assert record2.quality == "JJJJJJJJJJJJ"

        # Test third record
        record3 = records[2]
        assert record3.id == "seq3 short"
        assert record3.sequence == "ATCG"
        assert record3.quality == "AAAA"

    finally:
        fastq_file.unlink()


def test_read_fastq_convenience() -> None:
    """Test the read_fastq convenience function."""
    fastq_file = create_test_fastq()
    try:
        records = read_fastq(str(fastq_file))
        assert len(records) == 3
        assert records[0].id == "seq1 test sequence"
        assert records[0].sequence == "ATCGGATCCTAG"
        assert records[0].quality == "IIIIIIIIIIII"
    finally:
        fastq_file.unlink()


def test_fastq_record() -> None:
    """Test FastqRecord attributes and methods."""
    record = FastqRecord("test_id", "ATCG", "IIII")
    assert record.id == "test_id"
    assert record.sequence == "ATCG"
    assert record.quality == "IIII"

    # Test __repr__
    assert "FastqRecord" in repr(record)
    assert "test_id" in repr(record)
    assert "ATCG" in repr(record)
    assert "IIII" in repr(record)

    # Test equality
    record2 = FastqRecord("test_id", "ATCG", "IIII")
    record3 = FastqRecord("test_id", "ATCG", "JJJJ")  # Different quality
    assert record == record2
    assert record != record3


def test_gzip_compression() -> None:
    """Test reading gzip-compressed FASTQ files."""
    fastq_file = create_compressed_test_fastq('gzip')
    try:
        reader = FastqReader.from_file(str(fastq_file))
        records = list(reader)

        assert len(records) == 2
        assert records[0].id == "seq1 compressed"
        assert records[0].sequence == "ATCGGATCC"
        assert records[0].quality == "IIIIIIIII"
        assert records[1].id == "seq2 compressed too"
        assert records[1].sequence == "GGCCTTAA"
        assert records[1].quality == "JJJJJJJJ"
    finally:
        fastq_file.unlink()


def test_bzip2_compression() -> None:
    """Test reading bzip2-compressed FASTQ files."""
    fastq_file = create_compressed_test_fastq('bzip2')
    try:
        reader = FastqReader.from_file(str(fastq_file))
        records = list(reader)

        assert len(records) == 2
        assert records[0].id == "seq1 compressed"
        assert records[0].sequence == "ATCGGATCC"
        assert records[0].quality == "IIIIIIIII"
    finally:
        fastq_file.unlink()


def test_from_stdin_class_method() -> None:
    """Test the from_stdin class method."""
    reader = FastqReader.from_stdin()
    assert reader is not None
    assert hasattr(reader, '_reader')


def test_file_not_found() -> None:
    """Test error handling for missing files."""
    with pytest.raises(IOError):
        list(FastqReader.from_file("nonexistent_file.fastq"))

File: rust/tests/test_fastq.rs
// Tests for FASTQ parsing functionality
use prseq::fastq::{read_fastq, FastqReader};
use std::io::{Cursor, Write};
use tempfile::NamedTempFile;

#[test]
fn test_basic_fastq_reading() {
    let content = b"@seq1 test\nATCG\n+\nIIII\n@seq2 test\nGGCC\n+\nJJJJ\n";
    let cursor = Cursor::new(content);
    let mut reader = FastqReader::from_reader_with_capacity(cursor, 1024).unwrap();

    let record1 = reader.next().unwrap().unwrap();
    assert_eq!(record1.id, "seq1 test");
    assert_eq!(record1.sequence, "ATCG");
    assert_eq!(record1.quality, "IIII");

    let record2 = reader.next().unwrap().unwrap();
    assert_eq!(record2.id, "seq2 test");
    assert_eq!(record2.sequence, "GGCC");
    assert_eq!(record2.quality, "JJJJ");

    assert!(reader.next().is_none());
}

#[test]
fn test_multiline_fastq() {
    let content =
        b"@seq1 multiline\nATCG\nGCTA\n+seq1 multiline\nIIII\nJJJJ\n@seq2\nGG\nCC\n+\nKK\nLL\n";
    let cursor = Cursor::new(content);
    let mut reader = FastqReader::from_reader_with_capacity(cursor, 1024).unwrap();

    let record1 = reader.next().unwrap().unwrap();
    assert_eq!(record1.id, "seq1 multiline");
    assert_eq!(record1.sequence, "ATCGGCTA");
    assert_eq!(record1.quality, "IIIIJJJJ");

    let record2 = reader.next().unwrap().unwrap();
    assert_eq!(record2.id, "seq2");
    assert_eq!(record2.sequence, "GGCC");
    assert_eq!(record2.quality, "KKLL");

    assert!(reader.next().is_none());
}

#[test]
fn test_fastq_plus_line_validation() {
    // Test that '+' line with wrong ID fails
    let content = b"@seq1\nATCG\n+wrong_id\nIIII\n";
    let cursor = Cursor::new(content);
    let mut reader = FastqReader::from_reader_with_capacity(cursor, 1024).unwrap();

    let result = reader.next().unwrap();
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_string()
        .contains("does not match header ID"));
}

#[test]
fn test_fastq_length_mismatch() {
    // Test sequence/quality length mismatch
    let content = b"@seq1\nATCG\n+\nII\n"; // quality too short
    let cursor = Cursor::new(content);
    let mut reader = FastqReader::from_reader_with_capacity(cursor, 1024).unwrap();

    let result = reader.next().unwrap();
    assert!(result.is_err());
    let error_msg = result.unwrap_err().to_string();
    println!("Error message: {}", error_msg);
    assert!(
        error_msg.contains("Unexpected end of file")
            || error_msg.contains("does not match quality length")
    );
}

#[test]
fn test_fastq_file_reading() {
    let content = "@seq1 file test\nATCG\nGCTA\n+seq1 file test\nIIII\nJJJJ\n";
    let mut temp_file = NamedTempFile::new().unwrap();
    temp_file.write_all(content.as_bytes()).unwrap();

    let mut reader = FastqReader::from_file(temp_file.path()).unwrap();

    let record = reader.next().unwrap().unwrap();
    assert_eq!(record.id, "seq1 file test");
    assert_eq!(record.sequence, "ATCGGCTA");
    assert_eq!(record.quality, "IIIIJJJJ");

    assert!(reader.next().is_none());
}

#[test]
fn test_fastq_convenience_function() {
    let content = "@seq1\nATCG\n+\nIIII\n@seq2\nGGCC\n+\nJJJJ\n";
    let mut temp_file = NamedTempFile::new().unwrap();
    temp_file.write_all(content.as_bytes()).unwrap();

    let records = read_fastq(temp_file.path()).unwrap();

    assert_eq!(records.len(), 2);
    assert_eq!(records[0].id, "seq1");
    assert_eq!(records[0].sequence, "ATCG");
    assert_eq!(records[0].quality, "IIII");
    assert_eq!(records[1].sequence, "GGCC");
}

#[test]
fn test_invalid_fastq_start() {
    let content = b"INVALID LINE\nATCG\n+\nIIII\n";
    let cursor = Cursor::new(content);
    let mut reader = FastqReader::from_reader_with_capacity(cursor, 1024).unwrap();

    let result = reader.next().unwrap();
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_string()
        .contains("must start with '@'"));
}

#[test]
fn test_fastq_gzip_compression() {
    use flate2::write::GzEncoder;
    use flate2::Compression;

    let content = b"@seq1 compressed\nATCG\n+\nIIII\n@seq2 compressed\nGGCC\n+\nJJJJ\n";
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(content).unwrap();
    let compressed = encoder.finish().unwrap();

    let cursor = Cursor::new(compressed);
    let mut reader = FastqReader::from_reader_with_capacity(cursor, 1024).unwrap();

    let record1 = reader.next().unwrap().unwrap();
    assert_eq!(record1.id, "seq1 compressed");
    assert_eq!(record1.sequence, "ATCG");
    assert_eq!(record1.quality, "IIII");

    let record2 = reader.next().unwrap().unwrap();
    assert_eq!(record2.id, "seq2 compressed");
    assert_eq!(record2.sequence, "GGCC");
    assert_eq!(record2.quality, "JJJJ");

    assert!(reader.next().is_none());
}
